// @crayonai/stream/dist/jsonSchemaHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.templatesToResponseFormat = exports.zodToTemplateSchema = void 0;
const zod_to_json_schema_1 = require("zod-to-json-schema");
const textSchema_1 = require("./textSchema");
const zodToTemplateSchema = (schema, name, description) => {
    return {
        type: "object",
        description: description,
        properties: {
            type: {
                const: "template",
            },
            name: { const: name },
            templateProps: (0, zod_to_json_schema_1.zodToJsonSchema)(schema),
        },
        required: ["name", "templateProps", "type"],
        additionalProperties: false,
    };
};
exports.zodToTemplateSchema = zodToTemplateSchema;
const templatesToResponseFormat = (...templates) => {
    const templatesJsonSchema = {
        type: "object",
        properties: {
            response: {
                type: "array",
                items: {
                    oneOf: [
                        textSchema_1.TextResponseSchema,
                        ...templates.map((template) => (0, exports.zodToTemplateSchema)(template.schema, template.name, template.description)),
                    ],
                },
            },
        },
        required: ["response"],
        additionalProperties: false,
    };
    return {
        type: "json_schema",
        json_schema: {
            name: "json_schema",
            schema: templatesJsonSchema,
        },
    };
};
exports.templatesToResponseFormat = templatesToResponseFormat;
