// @crayonai/stream/dist/types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.MessageUpdate = exports.ContextUpdate = exports.ResponseTemplatePropsChunk = exports.ResponseTemplate = exports.TextChunk = exports.SSEType = void 0;
const eventsource_encoder_1 = require("eventsource-encoder");
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
var SSEType;
(function (SSEType) {
    SSEType["TextDelta"] = "text";
    SSEType["ResponseTemplate"] = "tpl";
    SSEType["ResponseTemplatePropsChunk"] = "tpl_props_chunk";
    SSEType["ContextAppend"] = "context_append";
    SSEType["MessageUpdate"] = "message_update";
    SSEType["Error"] = "error";
})(SSEType || (exports.SSEType = SSEType = {}));
class TextChunk {
    constructor(chunk) {
        this.chunk = chunk;
    }
    toSSEString() {
        return (0, eventsource_encoder_1.encode)({
            event: SSEType.TextDelta,
            data: this.chunk,
        });
    }
    static fromSSEData(data) {
        return new TextChunk(data);
    }
}
exports.TextChunk = TextChunk;
class ResponseTemplate {
    constructor(name, templateProps) {
        this.name = name;
        this.templateProps = templateProps;
    }
    toSSEString() {
        return (0, eventsource_encoder_1.encode)({
            event: SSEType.ResponseTemplate,
            data: JSON.stringify({ name: this.name, templateProps: this.templateProps }),
        });
    }
    static fromSSEData(data) {
        const { name, templateProps } = JSON.parse(data);
        (0, tiny_invariant_1.default)(name, "name is required in ResponseTemplate");
        return new ResponseTemplate(name, templateProps);
    }
}
exports.ResponseTemplate = ResponseTemplate;
class ResponseTemplatePropsChunk {
    constructor(chunk) {
        this.chunk = chunk;
    }
    toSSEString() {
        return (0, eventsource_encoder_1.encode)({
            event: SSEType.ResponseTemplatePropsChunk,
            data: this.chunk,
        });
    }
    static fromSSEData(data) {
        return new ResponseTemplatePropsChunk(data);
    }
}
exports.ResponseTemplatePropsChunk = ResponseTemplatePropsChunk;
class ContextUpdate {
    constructor(contextItem) {
        this.contextItem = contextItem;
    }
    toSSEString() {
        return (0, eventsource_encoder_1.encode)({
            event: SSEType.ContextAppend,
            data: JSON.stringify(this.contextItem),
        });
    }
    static fromSSEData(data) {
        return new ContextUpdate(JSON.parse(data));
    }
}
exports.ContextUpdate = ContextUpdate;
class MessageUpdate {
    constructor(id) {
        this.id = id;
    }
    toSSEString() {
        return (0, eventsource_encoder_1.encode)({
            event: SSEType.MessageUpdate,
            data: JSON.stringify({ id: this.id }),
        });
    }
    static fromSSEData(data) {
        const { id } = JSON.parse(data);
        (0, tiny_invariant_1.default)(id, "id is required in MessageUpdate");
        return new MessageUpdate(id);
    }
}
exports.MessageUpdate = MessageUpdate;
class Error {
    constructor(error) {
        this.error = error;
    }
    toSSEString() {
        return (0, eventsource_encoder_1.encode)({
            event: SSEType.Error,
            data: this.error,
        });
    }
    static fromSSEData(data) {
        return new Error(data);
    }
}
exports.Error = Error;
